File: plugins/moe/skills/running-an-iteration/scripts/validate_iteration_log.py
#!/usr/bin/env python3
"""Validate an iteration-log.md file.

Usage: validate_iteration_log.py <file>

Checks:
- Contains at least one ITER-NNNN section
- Each iteration section has Completed, Stories delivered, Tasks executed, Summary, Scenarios

Exit code: 0 on success, 1 on validation failure, 2 on invocation error.
"""
import re
import sys
from pathlib import Path


def validate_iteration_log(content: str) -> list[str]:
    """Validate an iteration-log.md file."""
    errors: list[str] = []

    iter_pattern = re.compile(r"^## ITER-(\d+)", re.MULTILINE)
    iters = list(iter_pattern.finditer(content))

    if not iters:
        errors.append("no iteration sections found (expected at least one '## ITER-NNNN')")
        return errors

    for idx, match in enumerate(iters):
        iter_id = f"ITER-{match.group(1)}"
        start = match.end()
        end = iters[idx + 1].start() if idx + 1 < len(iters) else len(content)
        section = content[start:end]

        for required in ("**Completed:**", "**Stories delivered:**",
                         "**Tasks executed:**", "**Scenarios:**", "**Summary:**"):
            if required not in section:
                errors.append(f"{iter_id}: missing required field {required}")

    return errors


def main() -> int:
    if len(sys.argv) != 2:
        print("usage: validate_iteration_log.py <file>", file=sys.stderr)
        return 2

    path = Path(sys.argv[1])
    if not path.exists():
        print(f"error: file not found: {path}", file=sys.stderr)
        return 2

    errors = validate_iteration_log(path.read_text())
    if errors:
        for err in errors:
            print(f"error: {err}", file=sys.stderr)
        return 1

    print(f"OK: {path}")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: packages/core/skills/running-an-iteration/scripts/validate_iteration_log.mjs
import { existsSync, readFileSync, realpathSync } from "node:fs";
import { resolve, sep } from "node:path";
import { fileURLToPath, pathToFileURL } from "node:url";

const ITERATION_PATTERN = /^## ITER-(\p{Nd}+)/gmu;
const REQUIRED_FIELDS = [
  "**Completed:**",
  "**Stories delivered:**",
  "**Tasks executed:**",
  "**Scenarios:**",
  "**Summary:**",
];

function pathParts(path, separatorPattern) {
  return path.split(separatorPattern).filter((part) => part && part !== ".");
}

function normalizeWindowsPathSpelling(path) {
  const windowsPath = path.replaceAll("/", "\\");
  const unc = windowsPath.match(/^\\\\([^\\]+)\\([^\\]+)(?:\\|$)/);
  if (unc) {
    const root = `\\\\${unc[1]}\\${unc[2]}\\`;
    const parts = pathParts(windowsPath.slice(unc[0].length), /\\+/);
    return parts.length > 0 ? `${root}${parts.join("\\")}` : root;
  }
  const drive = windowsPath.match(/^([A-Za-z]:)(\\?)/);
  if (drive) {
    const rooted = drive[2] === "\\";
    const parts = pathParts(windowsPath.slice(drive[1].length + (rooted ? 1 : 0)), /\\+/);
    if (rooted) return parts.length > 0 ? `${drive[1]}\\${parts.join("\\")}` : `${drive[1]}\\`;
    return parts.length > 0 ? `${drive[1]}${parts.join("\\")}` : drive[1];
  }
  const rooted = windowsPath.startsWith("\\");
  const parts = pathParts(windowsPath.slice(rooted ? 1 : 0), /\\+/);
  if (rooted) return parts.length > 0 ? `\\${parts.join("\\")}` : "\\";
  return parts.length > 0 ? parts.join("\\") : ".";
}

function normalizePathSpelling(path) {
  if (sep === "\\") return normalizeWindowsPathSpelling(path);
  const doubleSlashRoot = path.startsWith("//") && !path.startsWith("///");
  const root = doubleSlashRoot ? "//" : path.startsWith("/") ? "/" : "";
  const parts = pathParts(path, /\/+/);
  if (root) return parts.length > 0 ? `${root}${parts.join("/")}` : root;
  return parts.length > 0 ? parts.join("/") : ".";
}

export function validateIterationLog(content) {
  const errors = [];
  const iterations = [...content.matchAll(ITERATION_PATTERN)];
  if (iterations.length === 0) {
    return ["no iteration sections found (expected at least one '## ITER-NNNN')"];
  }

  for (const [index, iteration] of iterations.entries()) {
    const iterationId = `ITER-${iteration[1]}`;
    const start = iteration.index + iteration[0].length;
    const end = iterations[index + 1]?.index ?? content.length;
    const section = content.slice(start, end);
    for (const required of REQUIRED_FIELDS) {
      if (!section.includes(required)) {
        errors.push(`${iterationId}: missing required field ${required}`);
      }
    }
  }
  return errors;
}

export function main(args = process.argv.slice(2)) {
  if (args.length !== 1) {
    process.stderr.write("usage: validate_iteration_log.mjs <file>\n");
    return 2;
  }
  const path = normalizePathSpelling(args[0]);
  if (!existsSync(path)) {
    process.stderr.write(`error: file not found: ${path}\n`);
    return 2;
  }

  const errors = validateIterationLog(readFileSync(path, "utf8"));
  if (errors.length > 0) {
    for (const error of errors) process.stderr.write(`error: ${error}\n`);
    return 1;
  }
  process.stdout.write(`OK: ${path}\n`);
  return 0;
}

function isDirectEntry() {
  if (!process.argv[1]) return false;
  try {
    return realpathSync(resolve(process.argv[1])) === realpathSync(fileURLToPath(import.meta.url));
  } catch {
    return pathToFileURL(resolve(process.argv[1])).href === import.meta.url;
  }
}

if (isDirectEntry()) process.exitCode = main();
